crates/matrix-sdk/src/widget/machine/pending.rs

```diff
@@ -0,0 +1,173 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A wrapper around a hash map that tracks pending requests and makes sure
+//! that expired requests are removed.
+
+use std::time::{Duration, Instant};
+
+use indexmap::{map::Entry, IndexMap};
+use uuid::Uuid;
+
+/// Configuration of limits for the outgoing request handling.
+#[derive(Clone, Debug)]
+pub(crate) struct RequestLimits {
+    /// Maximum amount of unanswered (pending) requests that the client widget
+    /// API is going to process before starting to drop them. This ensures
+    /// that a buggy widget cannot force the client machine to consume memory
+    /// indefinitely.
+    pub(crate) max_pending_requests: usize,
+    /// For how long can the unanswered (pending) request stored in a map before
+    /// it is dropped. This ensures that requests that are not answered within
+    /// a ceratin amount of time, are dropped/cleaned up (considered as failed).
+    pub(crate) response_timeout: Duration,
+}
+
+/// A wrapper around a hash map that ensures that the request limits
+/// are taken into account (expired requests get cleaned up and that
+/// the hashmap remains limited to a certain amount of pending requests).
+pub(super) struct PendingRequests<T> {
+    requests: IndexMap<Uuid, Expirable<T>>,
+    limits: RequestLimits,
+}
+
+impl<T> PendingRequests<T> {
+    pub(super) fn new(limits: RequestLimits) -> Self {
+        Self { requests: IndexMap::with_capacity(limits.max_pending_requests), limits }
+    }
+
+    /// Inserts a new request into the map.
+    /// Returns `None` if the maximum allowed capacity is reached.
+    pub(super) fn insert(&mut self, key: Uuid, value: T) -> Option<&mut T> {
+        if self.requests.len() >= self.limits.max_pending_requests {
+            return None;
+        }
+
+        let Entry::Vacant(entry) = self.requests.entry(key) else {
+            panic!("uuid collision");
+        };
+
+        let expirable = Expirable::new(value, Instant::now() + self.limits.response_timeout);
+        let inserted = entry.insert(expirable);
+        Some(&mut inserted.value)
+    }
+
+    /// Extracts a request from the map based on its identifier.
+    /// Returns `None` if the value is not present or expired.
+    pub(super) fn extract(&mut self, key: &Uuid) -> Result<T, &'static str> {
+        let value = self.requests.remove(key).ok_or("Received response for an unknown request")?;
+        value.value().ok_or("Dropping response for an expired request")
```

Comment:
Do we have a strong reason to actually care about the expiry here?

Reply:
I thought that otherwise, we couldn't uphold an invariant that expired requests are not processed: imagine that we send a request, then we wait for 20 seconds and nothing comes back (no other requests are being sent or processed within this timeframe), then we suddenly receive a response. According to the expectations/spec for the timed-out requests, we must assume that this request is timed-out already, i.e. that the response is [meant to be] obsolete by this moment.

So by checking this inside a function, we can detect such cases and return an error (we log it in our case).

Reply:
No, I meant meant that I didn't expect expiry to be something meant to be enforced strictly, only as far as it helps defend against malicious widgets.

Reply:
Ah, so the limits for the maximum amount of pending requests are indeed just a safety mechanism against buggy or malicious widgets (it's just something that came into my mind when checking the code and realizing that we don't have any protection there). As for request timeouts (maximum amount of time to wait for the response before dropping it) - AFAIK this is something that is expected to be in place (though the timeout duration seems to be unspeced AFAIR the previous discussions with @toger5).

Reply:
I know its 10s in the js sdk (both widget and client)
https://github.com/matrix-org/matrix-widget-api/blob/9ab62e72c8b98087edcf1748d8c9c9f75631d030/src/transport/PostmessageTransport.ts#L44